### DISCUSSION TITLE
Copilot adapter for o1 models

### DISCUSSION THREAD

**Discussion by lucobellic** [olimorris/codecompanion.nvim#279] [Show and tell]
Using Copilot with o1 models through beta/limited access does not currently support all the default adapter options/requests. 

Since [codecompanion.nvim](https://github.com/olimorris/codecompanion.nvim) is extensible, it's really easy to create a new adapter to handle this. 

For anyone who may be interested, here is a small snippet to create an adapter for o1 models using Copilot:

> [!NOTE]
> Since `stream` is not supported, you will have to wait until the request is completely generated to see the output.

```lua
local M = {}

function M.get_adapter()
  local adapters = require('codecompanion.adapters')

  -- Extend the Copilot adapter with specific parameters and handlers for o1 models
  local adapter = adapters.extend('copilot', {
    opts = { stream = false }, -- Stream not supported
    schema = {
      model = {
        default = 'o1-preview',
        choices = {
          'o1-preview',
          'o1-mini',
        },
      },
    },
    handlers = {
      ---Handler to remove system prompt from messages
      ---@param self CodeCompanion.Adapter
      ---@param messages table
      form_messages = function(self, messages)
        return {
          messages = vim
            .iter(messages)
            :filter(function(message) return not (message.role and message.role == 'system') end)
            :totable(),
        }
      end,
    },
  })

  -- Remove unsupported settings from the adapter schema
  local unsupported_settings = { 'temperature', 'max_tokens', 'top_p', 'n' }
  vim.iter(unsupported_settings):each(function(setting) adapter.schema[setting] = nil end)

  adapter.name = 'copilot_o1'
  return adapter
end

return M
```

```lua
{
  'olimorris/codecompanion.nvim',
  ...
  opts = {
    ...
    adapters = {
      copilot_o1 = require('[path to adapter]').get_adapter,
    },
  }
}
```

**Comment by olimorris** [comment]
This is awesome! Thanks for sharing. How have you found the o1 model's performance so far?

**Reply by olimorris** [reply]
I get the following:

```
[ERROR] 2024-10-01 11:39:36
Error: The requested model is not supported.
```

I wonder if it's region specific.

**Reply by S1M0N38** [reply]
you have to join the waiting list and they have to give you access.

**Reply by lucobellic** [reply]
I didn't test the models extensively, but from my current use, I found o1-preview to always correctly answer my code questions or refactoring requests and to have a better "understanding" of the code and what I am trying to achieve.

Overall, it's better compared to Copilot with GPT-4o. The main advantage I found is that I don't have to interact as much as with other models, requiring less description of what I want and fewer corrections when the model does not perform exactly as I would expect.

**Reply by serpro69** [reply]
Thanks a lot for sharing. I'm going to try this to see if it improves my experience.

I'm using copilot with gpt-4o currently and I find it extremely "dumb" and useless more often than not. Even a regular free chatgpt thing usually gives me better answers to the same question :/ Copilot adapter just apologizes for confusion all the time  :rofl: and provides unchanged code and whatnot.
Not sure if I'm doing smth wrong with the setup, this is pretty much my plugin config:
```lua
{
      "olimorris/codecompanion.nvim",
      enabled = true,
      event = "BufReadPost",
      dependencies = {
        "nvim-lua/plenary.nvim",
        "nvim-treesitter/nvim-treesitter",
        "nvim-telescope/telescope.nvim", -- Optional
        {
          "stevearc/dressing.nvim", -- Optional: Improves the default Neovim UI
          opts = {},
        },
      },
      config = function()
        require("codecompanion").setup {
          adapters = {
            copilot = function()
              return require("codecompanion.adapters").extend("copilot", {
                schema = {
                  model = {
                    default = "gpt-4o", -- tried to change from default and also tried gpt4-turbo - neither gave a noticeable improvement to response quality
                  },
                },
              })
            end,
          },
          strategies = {
            chat = {
              adapter = "copilot",
            },
            inline = {
              adapter = "copilot",
            },
            agent = {
              adapter = "copilot",
            },
          },
          display = {
            chat = {
              window = {
                layout = "vertical", -- float|vertical|horizontal|buffer
              },
            },
            inline = {
              diff = {
                hl_groups = {
                  added = "DiffAdd",
                },
              },
            },
          },
          opts = {
            log_level = "DEBUG",
          },
        }
      end,
      init = function()
        vim.cmd [[cab cc CodeCompanion]]
        require("legendary").keymaps {
          {
            itemgroup = "CodeCompanion",
            icon = "",
            description = "Use the power of AI...",
            keymaps = {
              {
                "<C-a>",
                "<cmd>CodeCompanionActions<CR>",
                description = "Open the CodeCompanion action picker",
                mode = { "n", "v" },
              },
              {
                "<LocalLeader>a",
                "<cmd>CodeCompanionToggle<CR>",
                description = "Open CodeCompanion chat prompt",
                mode = { "n", "v" },
              },
              {
                "ga",
                "<cmd>CodeCompanionAdd<CR>",
                description = "Add selected text to CodeCompanion",
                mode = { "n", "v" },
              },
            },
          },
        }
      end,
    }
```

**Reply by lucobellic** [reply]
@serpro69 I haven't recently reused copilot with default options, but I can reproduce what you're describing. 
For some reason you have to explicitly set the stream to false (`opts = {stream = false}`) to produces responses as you could expect from gpt-4o model instead of the "dumb" one.

**Reply by serpro69** [reply]
Oh, really? Thanks for the tip @lucobellic , I'll try that.
I really like the plugin so far, but using copilot model with defaults has been nothing but "pain" :D 
Maybe this could be documented in copilot adapter readme?

**Reply by PyDataBlog** [reply]
> This is awesome! Thanks for sharing. How have you found the o1 model's performance so far?

Now copilot supports all the big commercial models. How can we select a model for this adapter? ties into #387

**Comment by olimorris** [comment]
@lucobellic I'm in the process of tweaking the adapter and http client. You'll need to change:

```lua
parameters = { stream = false }, -- Stream not supported
```

to:

```lua
opts = { stream = false }, -- Stream not supported
```
